File: Java/Slidingwindow/Count distinct elements in every window.java
//https://www.geeksforgeeks.org/problems/count-distinct-elements-in-every-window/1
class Solution {
    ArrayList<Integer> countDistinct(int arr[], int k) {
        // code here
        ArrayList<Integer> result = new ArrayList<>();
        Map<Integer, Integer> map = new HashMap<>();
        int i = 0, j = 0;

        while (j < arr.length) {
            map.put(arr[j], map.getOrDefault(arr[j], 0) + 1);

            if (j - i + 1 == k) {
                result.add(map.size());

                // Slide the window
                map.put(arr[i], map.get(arr[i]) - 1);
                if (map.get(arr[i]) == 0) {
                    map.remove(arr[i]);
                }
                i++;
            }
            j++;
        }
        return result;
    }
}

File: Python/Slidingwindow/Count distinct elements in every window.py
//https://www.geeksforgeeks.org/problems/count-distinct-elements-in-every-window/1
class Solution:
    def countDistinct(self, arr, k):
        # Code here
        result = []
        freq = {}
        i = 0

        for j in range(len(arr)):
            freq[arr[j]] = freq.get(arr[j], 0) + 1

            if j - i + 1 == k:
                result.append(len(freq))

                # Slide the window
                freq[arr[i]] -= 1
                if freq[arr[i]] == 0:
                    del freq[arr[i]]
                i += 1

        return result
